utils/core/src/main/kotlin/NetRcAuthenticator.kt

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (C) 2020-2021 Bosch.IO GmbH
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.utils.core
+
+import java.net.Authenticator
+import java.net.PasswordAuthentication
+
+import org.ossreviewtoolkit.utils.common.Os
+import org.ossreviewtoolkit.utils.common.nextOrNull
+
+/**
+ * A simple non-caching authenticator that reads credentials from .netrc-style files.
+ */
+class NetRcAuthenticator : Authenticator() {
+    // TODO: Add support for ".authinfo" files (which use the same syntax as .netrc files) once Git.kt does not call the
+    //       Git CLI anymore which only supports ".netrc" (and "_netrc") files.
+    private val netrcFileNames = listOf(".netrc", "_netrc")
+
+    override fun getPasswordAuthentication(): PasswordAuthentication? {
+        netrcFileNames.forEach { name ->
+            val netrcFile = Os.userHomeDirectory.resolve(name)
+            if (netrcFile.isFile) {
+                log.debug { "Parsing '$netrcFile' for machine '$requestingHost'." }
+
+                // Read the file on each function call to be up-to-date with any changes.
+                val netrcText = netrcFile.readText()
+
+                return getNetrcAuthentication(netrcText, requestingHost)?.let { return it }
+            }
+        }
+
+        return super.getPasswordAuthentication()
+    }
+}
+
+/**
+ * Parse the [contents] of a [.netrc](https://www.gnu.org/software/inetutils/manual/html_node/The-_002enetrc-file.html)
+ * file for a login / password matching [machine].
+ */
+internal fun getNetrcAuthentication(contents: String, machine: String): PasswordAuthentication? {
```

Comment:
Just wondering: Instead of parsing the netrc files for every machine, wouldn't it be more efficient to read them once and extract the credentials for all machines?

Reply:
My thinking was that for this non-caching low-level authenticator we might want to be able to consider changes to `.netrc` files during the runtime of the application (mainly for on-the-fly debugging). If that's not desired, a higher level authenticator can implement caching, like the `OrtAuthenticator` does.